### PR TITLE
balloon_hotplug: hot-unplug balloon device when balloon service is running

### PR DESCRIPTION
**Pull request by lijinlijin**
hot-unplug balloon device while balloon service is enabled.

ID:1456109

Signed-off-by: Li Jin <lijin@redhat.com>

### CODE REVIEW

**Comment by lijinlijin** [comment]
based on avocado-framework/avocado-vt#1422

**Comment by lijinlijin** [comment]
PASS 1-smp_2.4096m.repeat1.run_test.Host_RHEL.m7.u5.qcow2.virtio_blk.up.virtio_net.Guest.Win10.x86_64.io-github-autotest-qemu.balloon_hotplug.with_balloon_service

File: qemu/tests/cfg/balloon_hotplug.cfg [hunk 1]
```diff
@@ -13,6 +13,18 @@
     balloon_type_evict = evict
     balloon_type_enlarge = enlarge
     balloon_device = virtio-balloon-pci
+    Windows:
+        cdroms = "virtio"
```
Comment (xiagao) [inline qemu/tests/cfg/balloon_hotplug.cfg]: It's better to use cdroms += " virtio"

File: qemu/tests/cfg/balloon_hotplug.cfg [hunk 2]
```diff
@@ -13,6 +13,18 @@
...
+        # Please change following cdrom_virtio to the right version
+        # if necessary
+        cdrom_virtio = isos/windows/virtio-win.iso
+        # balloon service related cmd, which will be covered by different
+        # guest's paths in guest-os cfg.
+        install_balloon_service = "%s:\Balloon\GUEST_OS\amd64\blnsvr.exe -i"
+        uninstall_balloon_service = "%s:\Balloon\GUEST_OS\amd64\blnsvr.exe -u"
+        status_balloon_service = "%s:\Balloon\GUEST_OS\amd64\blnsvr.exe status"
+        run_balloon_service = "%s:\Balloon\GUEST_OS\amd64\blnsvr.exe -r"
+        stop_balloon_service = "%s:\Balloon\GUEST_OS\amd64\blnsvr.exe -s"
```
Comment (xiagao) [inline qemu/tests/cfg/balloon_hotplug.cfg]: If it's better to move those cfg to with_balloon_service as they are only available for balloon service ?

File: qemu/tests/cfg/balloon_hotplug.cfg (hunk 1 above)
Comment (lijinlijin) [reply qemu/tests/cfg/balloon_hotplug.cfg]: Is there any reason?

File: qemu/tests/cfg/balloon_hotplug.cfg (hunk 2 above)
Comment (lijinlijin) [reply qemu/tests/cfg/balloon_hotplug.cfg]: I think it should be ok, any new variant can reuse these parameters in future, no need to rewrite again

File: qemu/tests/balloon_hotplug.py [hunk 3]
```diff
@@ -36,6 +39,19 @@ def run_pm_test(pm_test, plug_type):
...
+        :param session: shell Object
+        """
+        if params['os_type'] == 'windows':
+            balloon_test = BallooningTestWin(test, params, env)
+            balloon_test.configure_balloon_service(session)
+
+            output = balloon_test.operate_balloon_service(session, "status")
+            if not re.search("running", output.lower(), re.M):
+                test.error("Ballooon service status is not running")
+
```
Comment (huangyum) [inline qemu/tests/balloon_hotplug.py]: IIUC, if guest didn't install balloon service at beginning, L49 will install it. Then would it be running automatically after installation ? Or we need to call operate_balloon_service(session, "run") before L51?

File: qemu/tests/balloon_hotplug.py [hunk 4]
```diff
@@ -70,6 +87,11 @@ def run_pm_test(pm_test, plug_type):
         vm.params["balloon_dev_devid"] = "balloon%d" % idx
         vm.params["balloon_dev_add_bus"] = "yes"
 
+        if params.get("blnsrv_enable", "no") == "yes":
+            error_context.context("Install and check balloon service in windows"
+                                  "guest", logging.info)
+            enable_balloon_service(session)
```
Comment (huangyum) [inline qemu/tests/balloon_hotplug.py]: Do we need to close the session if we will not use it anymore?

File: qemu/tests/balloon_hotplug.py (hunk 3 above)
Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: After installation, it will be in running status automatically

File: qemu/tests/balloon_hotplug.py (hunk 4 above)
Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: Sure, will modify it.
Thanks

Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: Modified.

File: qemu/tests/balloon_hotplug.py [hunk 5]
```diff
@@ -14,11 +16,12 @@ def run(test, params, env):
 
     1) Boot up guest w/o balloon device.
     2) Hoplug balloon device and check hotplug successfully or not.
-    3) Do memory balloon.
-    4) Reboot/shutdown/migrate guest after hotplug balloon device(option)
-    5) Do memory balloon after guest reboot(option)
-    6) Unplug balloon device and check unplug successfully or not.
-    7) Reboot/shutdown/migrate guest after unplug balloon device(option)
+    3) Install balloon service and check its status in windows guests(option).
```
Comment (huangyum) [inline qemu/tests/balloon_hotplug.py]: Do we need to verify the driver or something? I'm not sure. Other parts look good to me.

Comment (lijinlijin) [reply qemu/tests/balloon_hotplug.py]: I add driver check step in this step as well.
Thanks.

**Comment by lijinlijin** [comment]
@xiagao @huangyum 
Any new comment?

**Comment by huangyum** [comment]
LGTM,
Acked-by: Yumei Huang <yuhuang@redhat.com>

File: qemu/tests/balloon_hotplug.py [hunk 6]
```diff
@@ -70,6 +91,11 @@ def run_pm_test(pm_test, plug_type):
         vm.params["balloon_dev_devid"] = "balloon%d" % idx
         vm.params["balloon_dev_add_bus"] = "yes"
 
+        if params.get("blnsrv_enable", "no") == "yes":
```
Comment (luckyh) [inline qemu/tests/balloon_hotplug.py]: Do we need to run this test with `blnsrv_enable = no` on Windows? If no, we do not need this param, and the thing would be like:
```python
def enable_balloon_service():
    if params['os_type'] != 'windows':
        return
    error_context.context("Install and check balloon service in windows"
                          "guest", logging.info)
    session = vm.wait_for_login()
    ...


...
enable_balloon_service()
...
```

File: qemu/tests/balloon_hotplug.py [hunk 7]
```diff
@@ -36,6 +39,24 @@ def run_pm_test(pm_test, plug_type):
...
+            session = vm.wait_for_login()
+            driver_name = params.get("driver_name", "balloon")
+            session = utils_test.qemu.windrv_check_running_verifier(session,
+                                                                    vm, test,
+                                                                    driver_name)
+            balloon_test = BallooningTestWin(test, params, env)
+            balloon_test.configure_balloon_service(session)
+
+            output = balloon_test.operate_balloon_service(session, "status")
+            if not re.search("running", output.lower(), re.M):
```
Comment (luckyh) [inline qemu/tests/balloon_hotplug.py]: Suggest `"running"` -> `r"running"`

**Comment by lijinlijin** [comment]
@luckyh 
Modified.
Run with latest code, all pass:
 (1/7) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.balloon_hotplug: PASS (54068.36 s)
 (2/7) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.balloon_hotplug.with_reboot: PASS (1276.72 s)
 (3/7) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.balloon_hotplug.with_system_reset: PASS (1141.64 s)
 (4/7) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_unplug: PASS (658.59 s)
 (5/7) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.balloon_hotplug.with_shutdown_after_plug: PASS (794.50 s)
 (6/7) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_hotplug: PASS (748.82 s)
 (7/7) Host_RHEL.m7.u5.qcow2.virtio_scsi.up.virtio_net.Guest.Win10.i386.io-github-autotest-qemu.balloon_hotplug.with_migrate.migrate_after_unplug: PASS (685.47 s)

**Pull request merged**